src/util/expr_iterator.h

```diff
@@ -0,0 +1,228 @@
+// to access non-const exprt reference but is an expensive operation
+
+// Naive depth-first-search iterator with .mutate method
+class depth_iteratort;
+// Naive depth-first-search iterator
+class const_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance and a .mutate method
+// Slower iteration than naive version. Very expensive copy.
+class unique_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance
+// Slower iteration than naive version. Very expensive copy.
+class const_unique_depth_iteratort;
+
+// Depth first search iterator - iterates over supplied expression
+// and all its operands recursively.
+// Base class using CRTP
+
+template<typename depth_iterator_t>
+class depth_iterator_baset:
+public std::iterator<std::forward_iterator_tag, const exprt, void>
+{
+public:
+  bool operator==(const depth_iterator_t &other) const
+  {
+    return m_stack.size()==other.m_stack.size() &&
+           std::equal(m_stack.begin(),
+                      m_stack.end(),
+                      other.m_stack.begin());
+  }
+  bool operator!=(const depth_iterator_t &other) const
+  {
+    return !(*this==other);
+  }
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+        {
+          break;
+        }
+      }
+      else if(this->downcast().push_expr(*m_stack.back().it))
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+  const exprt &operator*() const
+  {
+    return m_stack.back().expr.get();
+  }
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+protected:
+  // Create end iterator
+  depth_iterator_baset() { }
+  // Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+  typedef typename std::vector<exprt>::const_iterator operands_iteratort;
+
+  struct expr_statet final
+  {
+    expr_statet(
+      const exprt &expr,
+      operands_iteratort it,
+      operands_iteratort end):
+      expr(expr), it(it), end(end) { }
+    expr_statet(const expr_statet &other):
+      expr(other.expr), it(other.it), end(other.end) { }
+    std::reference_wrapper<const exprt> expr;
+    // Cache to avoid unnecessary mutable operands() calls
+    std::vector<exprt> *operands=nullptr;
+    operands_iteratort it;
+    const operands_iteratort end;
+    bool operator==(const expr_statet &other) const
+    {
+      return it==other.it;
+    }
+  };
+  // Obtain non-const exprt reference. Breaks the tree as a side
+  // effect. To be called only if a non-const exprt was passed into
+  // the constructor Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt&>(m_stack.front().expr.get()));
```

Comment:
Where does this code actually build a copy? You need to create a local copy here, and at the very end swap the expressions. As the (swapped!) local copy goes out of scope it will decrement the reference count, and may get destroyed (if it is no longer used elsewhere). Only upon doing the swap should you need to do a `const_cast`.

Reconstruct the stack based on that new, local expression. Be aware, however, that the depth iterator object becomes unusable once the expression returned by `mutate` is modified (because the iterators stored in `expr_statet` may be invalidated). This may limit this usefulness of this iterator.